ui/packages/atlasmap/src/UI/AddFieldTypeahead.tsx

```diff
@@ -0,0 +1,87 @@
+import React, { FunctionComponent, ChangeEvent, useCallback } from "react";
+import {
+  Select,
+  SelectOption,
+  Text,
+  TextContent,
+  TextVariants,
+  SelectProps,
+} from "@patternfly/react-core";
+import { useToggle } from "./useToggle";
+
+export interface IAddFieldTypeaheadField {
+  label: string;
+  info?: string;
+  onAdd: () => void;
+}
+
+export interface IAddFieldTypeaheadProps {
+  fields: IAddFieldTypeaheadField[];
+  ariaLabelTypeAhead: string;
+  placeholderText: string;
+}
+
+export const AddFieldTypeahead: FunctionComponent<IAddFieldTypeaheadProps> = ({
+  fields,
+  ariaLabelTypeAhead,
+  placeholderText,
+  ...props
+}) => {
+  const { state, toggle, toggleOff } = useToggle(false);
+
+  const renderOptions = (fields: IAddFieldTypeaheadField[]) => {
+    return fields.map((f, index) => (
+      <SelectOption
+        key={index}
+        value={{
+          ...f,
+          toString: () => f.label,
+          compareTo: (c) =>
+            f.label.localeCompare((c as IAddFieldTypeaheadField).label) === 0,
+        }}
+        data-testid={`add-field-option-${f.label}`}
+      >
+        <TextContent>
+          {f.label}
+          {f.info && <Text component={TextVariants.small}>{f.info}</Text>}
+        </TextContent>
+      </SelectOption>
+    ));
+  };
+
+  const filterFields = useCallback(
+    (e: ChangeEvent<HTMLInputElement>) => {
+      try {
+        const searchValueRX = new RegExp(e.target.value, "i");
+        return renderOptions(fields.filter((f) => searchValueRX.test(f.label)));
+      } catch (err) {}
+      return renderOptions(fields);
+    },
+    [fields],
+  );
+
+  const onSelect: SelectProps["onSelect"] = useCallback(
+    (_e, f) => {
+      (f as IAddFieldTypeaheadField).onAdd();
+      toggleOff();
+    },
+    [toggleOff],
+  );
+
+  return (
+    <div {...props}>
+      <Select
+        variant={"typeahead"}
+        ariaLabelTypeAhead={ariaLabelTypeAhead}
+        onToggle={toggle}
+        isExpanded={state}
+        placeholderText={placeholderText}
+        onFilter={filterFields}
+        onSelect={onSelect}
+        maxHeight={300}
+      >
+        {renderOptions(fields)}
```

Comment:
Need to add properties and constants as selectable fields

Reply:
Ops, right, thanks!

Reply:
https://github.com/atlasmap/atlasmap/pull/2071/files#diff-ed5af85d7b40907087a3ae8f3ea5bd57R118-R124